gtsam/nonlinear/FunctorizedFactor.h

```diff
@@ -0,0 +1,145 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file FunctorizedFactor.h
+ * @date May 31, 2020
+ * @author Varun Agrawal
+ **/
+
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+#include <cmath>
+
+namespace gtsam {
+
+/**
+ * Factor which evaluates functor and uses the result to compute
+ * error on provided measurement.
+ * The provided FUNCTOR should provide two type aliases: `argument_type` which
```

Comment:
Now I’m wondering whether this is just any `std::function`? And whether we could use it with a lambda? https://en.cppreference.com/w/cpp/utility/functional/function

I propose you add two unit tests that exercise those cases and we can see if and how it leads to a change in your requirements. It could also be much simpler to just say that the Functor has to satisfy the standard function concept

In fact, reading more, we probably have to:

https://stackoverflow.com/questions/35907454/why-stdfunctionargument-type-has-been-deprecated

“Now that we have decltype that can be used to deduce a callable's return type, variadic templates and perfect forwarding to forward arbitrary number of arguments to functions, etc. the pre-C++11 mechanisms are way too cumbersome to use.”


https://en.cppreference.com/w/cpp/language/decltype

The following post seems to hint that we might not *want* to know the argument types though: https://stackoverflow.com/questions/6512019/can-we-get-the-type-of-a-lambda-argument

Reply:
I actually thought about this at first and had some initial difficulty which I don't remember right now. Then again, my understanding of functionals in C++ has improved quite a bit since then so I'll tackle this today.

Reply:
So I've not had a lot of success with this without over-complicating the code. Something to consider would be just removing the need for `argument_type` and `return_type` from the `FUNCTOR` and making it match the traits of `std::function` so that they can be used interchangeably.

Reply:
Yes. And it should also do lambdas. Did you make the two unit tests I proposed?

Reply:
I did, but I've been having trouble getting the code to compile.